src/perf/event_provider.cpp

```diff
@@ -0,0 +1,384 @@
+
+                if (event_is_openable(ev))
+                {
+                    map.emplace(name_fmt.str(), ev);
+                }
+            }
+        }
+    }
+}
+
+static std::uint64_t parse_bitmask(const std::string& format)
+{
+    enum format_regex_groups
+    {
+        WHOLE_MATCH,
+        BIT_BEGIN,
+        BIT_END,
+    };
+
+    std::uint64_t mask = 0x0;
+
+    static const std::regex bit_mask_regex(R"((\d+)?(?:-(\d+)))");
+    static const std::sregex_iterator end;
+    std::smatch bit_mask_match;
+    for (std::sregex_iterator i = { format.begin(), format.end(), bit_mask_regex }; i != end; i++)
+    {
+        const auto& match = *i;
+        int start = std::stoi(match[BIT_BEGIN]);
+        int end = (match[BIT_END].length() == 0) ? start : std::stoi(match[BIT_END]);
+
+        const auto len = (end + 1) - start;
+        if (start < 0 || end > 63 || len > 64)
+        {
+            throw EventProvider::InvalidEvent("invalid config mask");
+        }
+
+        /* set `len` bits and shift them to where they should start.
+         * 4-bit example: format "1-3" produces mask 0x1110.
+         *    start := 1, end := 3
+         *    len  := 3 + 1 - 1 = 3
+         *    bits := (1 << 3) - 1 = 0b1000 - 1 = 0b0111
+         *    mask := 0b0111 << 1 = 0b1110
+         * */
+        const auto bits = (len == 64) ? ~0x0UL : (1 << len) - 1;
+        mask |= bits << start;
+    }
+    Log::debug() << std::showbase << std::hex << "config mask: " << format << " = " << mask
+                 << std::dec << std::noshowbase;
+    return mask;
+}
+
+static constexpr std::uint64_t apply_mask(std::uint64_t value, std::uint64_t mask)
+{
+    std::uint64_t res = 0;
+    for (int mask_bit = 0, value_bit = 0; mask_bit < 64; mask_bit++)
+    {
+        if (mask & (1 << mask_bit))
+        {
+            res |= ((value >> value_bit) & 0x1) << mask_bit;
+            value_bit++;
+        }
+    }
+    return res;
+}
+
+static void event_description_update(platform::CounterDescription& event, std::uint64_t value,
+                                     const std::string& format)
+{
+    static constexpr auto npos = std::string::npos;
+    const auto colon = format.find_first_of(':');
+    if (colon == npos)
+    {
+        throw EventProvider::InvalidEvent("invalid format description: missing colon");
+    }
+
+    const std::string target_config = format.substr(0, colon);
+    const auto mask = parse_bitmask(format.substr(colon + 1));
+
+    if (target_config == "config")
+    {
+        event.config |= apply_mask(value, mask);
+    }
+
+    if (target_config == "config1")
+    {
+        event.config1 |= apply_mask(value, mask);
+    }
+}
+
+const platform::CounterDescription sysfs_read_event(const std::string& ev_desc)
+{
+    namespace fs = boost::filesystem;
+    static constexpr auto npos = std::string::npos;
+    const auto slash_pos = ev_desc.find_first_of('/');
```

Comment:
perf requires a trailing `/`, we should at least allow that. It might be helpful to require it for disambiguation.